File: app/ios/MedNote/Services/AIService.swift
import Foundation

struct AIService {
    struct PipelineResult {
        var transcript: String
        var transcriptLines: [String]
        var summary: String
        var extractedItems: [ExtractedItem]
        var translatedTranscript: [String]?
    }

    static func runPipeline(
        audioURL: URL,
        groqKey: String,
        geminiKey: String,
        userLanguage: AppLanguage,
        recordLanguage: AppLanguage,
        durationSeconds: Int
    ) async throws -> PipelineResult {
        let gKey = groqKey.trimmingCharacters(in: .whitespaces)
        let cKey = geminiKey.trimmingCharacters(in: .whitespaces)
        let userLangName = userLanguage.label
        let origLang = recordLanguage.rawValue

        var originalText = "[No transcript — add a Groq API key in Settings]"
        if !gKey.isEmpty {
            if let whisper = try? await groqWhisper(fileURL: audioURL, apiKey: gKey, language: recordLanguage.whisperCode) {
                originalText = whisper
            }
        }

        let lines = splitTranscriptLines(originalText)
        let shouldTranslate = userLanguage != recordLanguage && !originalText.hasPrefix("[")

        async let summaryTask = callGemini(
            system: "Medical appointment summarizer. Warm, concise, plain language.",
            message: """
            Brief summary in \(userLangName). Headings: 📋 Findings 💊 Medications ⚠️ Watch for 📅 Next steps

            Transcript:
            \(originalText)

            Write entirely in \(userLangName).
            """,
            geminiKey: cKey,
            groqKey: gKey
        )

        async let extractTask = callGemini(
            system: "Medical data extractor. Return ONLY a valid JSON array, no markdown, no other text.",
            message: """
            Extract actionable items from the transcript. Respond ONLY with a valid JSON array (no markdown, no extra text), each item: {"type":"medication"|"appointment"|"action","text":"description in \(userLangName)","time":"time/date","suggestedHour":8}

            IMPORTANT: All "text" values MUST be written in \(userLangName).

            Transcript:
            \(originalText)
            """,
            geminiKey: cKey,
            groqKey: gKey
        )

        async let translateTask: [String]? = {
            guard shouldTranslate else { return nil }
            let result = await callGemini(
                system: "Medical transcript translator. Keep drug names, dosages, and numbers accurate. Return ONLY the translated transcript.",
                message: "Translate into \(userLangName):\n\n\(originalText)",
                geminiKey: cKey,
                groqKey: gKey
            )
            guard let result else { return nil }
            return splitTranscriptLines(result)
        }()

        let summary = await summaryTask ?? "AI summary could not be generated — add a Groq or Gemini key in Settings."
        let extractRaw = await extractTask
        let translated = await translateTask
        let items = parseExtractedItems(extractRaw, language: userLanguage)

        _ = durationSeconds
        return PipelineResult(
            transcript: originalText,
            transcriptLines: lines.isEmpty ? [originalText] : lines,
            summary: summary,
            extractedItems: items,
            translatedTranscript: translated
        )
    }

    // MARK: - Groq Whisper

    private static func groqWhisper(fileURL: URL, apiKey: String, language: String?) async throws -> String? {
        let data = try Data(contentsOf: fileURL)
        let boundary = "Boundary-\(UUID().uuidString)"
        var body = Data()
        let ext = fileURL.pathExtension.isEmpty ? "m4a" : fileURL.pathExtension
        let mime = ext == "wav" ? "audio/wav" : "audio/mp4"

        func append(_ string: String) {
            body.append(Data(string.utf8))
        }

        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"file\"; filename=\"audio.\(ext)\"\r\n")
        append("Content-Type: \(mime)\r\n\r\n")
        body.append(data)
        append("\r\n")
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"model\"\r\n\r\n")
        append("whisper-large-v3-turbo\r\n")
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"response_format\"\r\n\r\n")
        append("verbose_json\r\n")
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"temperature\"\r\n\r\n")
        append("0\r\n")
        if let language {
            append("--\(boundary)\r\n")
            append("Content-Disposition: form-data; name=\"language\"\r\n\r\n")
            append("\(language)\r\n")
        }
        append("--\(boundary)--\r\n")

        var request = URLRequest(url: URL(string: "https://api.groq.com/openai/v1/audio/transcriptions")!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let (responseData, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
        struct WhisperResponse: Decodable { let text: String? }
        return try JSONDecoder().decode(WhisperResponse.self, from: responseData).text
    }

    // MARK: - Gemini / Groq chat

    private static func callGemini(system: String, message: String, geminiKey: String, groqKey: String) async -> String? {
        let prompt = "\(system)\n\n\(message)"
        let gKey = geminiKey.trimmingCharacters(in: .whitespaces)
        let qKey = groqKey.trimmingCharacters(in: .whitespaces)

        if !gKey.isEmpty {
            for model in ["gemini-2.5-flash", "gemini-2.5-flash-lite", "gemini-flash-latest"] {
                if let text = await geminiGenerate(prompt: prompt, apiKey: gKey, model: model) {
                    return text
                }
            }
        }

        if !qKey.isEmpty {
            return await groqChat(prompt: prompt, apiKey: qKey)
        }
        return nil
    }

    private static func geminiGenerate(prompt: String, apiKey: String, model: String) async -> String? {
        guard let url = URL(string: "https://generativelanguage.googleapis.com/v1beta/models/\(model):generateContent?key=\(apiKey)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = [
            "contents": [["role": "user", "parts": [["text": prompt]]]],
            "generationConfig": ["maxOutputTokens": 1500, "temperature": 0.3]
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
        struct GeminiResponse: Decodable {
            struct Candidate: Decodable {
                struct Content: Decodable {
                    struct Part: Decodable { let text: String? }
                    let parts: [Part]?
                }
                let content: Content?
            }
            struct ErrorBody: Decodable { struct Err: Decodable { let message: String? }; let error: Err? }
            let candidates: [Candidate]?
            let error: ErrorBody.Err?
        }
        guard let decoded = try? JSONDecoder().decode(GeminiResponse.self, from: data),
              decoded.error == nil else { return nil }
        return decoded.candidates?.first?.content?.parts?.first?.text
    }

    private static func groqChat(prompt: String, apiKey: String) async -> String? {
        var request = URLRequest(url: URL(string: "https://api.groq.com/openai/v1/chat/completions")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        let body: [String: Any] = [
            "model": "llama-3.3-70b-versatile",
            "messages": [["role": "user", "content": prompt]],
            "max_tokens": 1500,
            "temperature": 0.3
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
        struct GroqResponse: Decodable {
            struct Choice: Decodable {
                struct Message: Decodable { let content: String? }
                let message: Message?
            }
            let choices: [Choice]?
        }
        return (try? JSONDecoder().decode(GroqResponse.self, from: data))?.choices?.first?.message?.content
    }

    // MARK: - Helpers

    private static func splitTranscriptLines(_ text: String) -> [String] {
        guard !text.isEmpty else { return [] }
        var lines: [String] = []
        text.enumerateSubstrings(in: text.startIndex..., options: [.bySentences, .substringNotRequired]) { substring, _, _, _ in
            if let s = substring?.trimmingCharacters(in: .whitespacesAndNewlines), s.count > 1 {
                lines.append(s)
            }
        }
        if lines.isEmpty {
            return text.components(separatedBy: .newlines).map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        }
        return lines
    }

    private static func parseExtractedItems(_ raw: String?, language: AppLanguage) -> [ExtractedItem] {
        guard var raw else { return [] }
        raw = raw.replacingOccurrences(of: "```json", with: "").replacingOccurrences(of: "```", with: "")
        if let start = raw.firstIndex(of: "["), let end = raw.lastIndex(of: "]") {
            raw = String(raw[start...end])
        }
        struct RawItem: Decodable {
            let type: String?
            let text: String?
            let time: String?
            let suggestedHour: Int?
        }
        guard let data = raw.data(using: .utf8),
              let parsed = try? JSONDecoder().decode([RawItem].self, from: data) else { return [] }
        return parsed.enumerated().compactMap { idx, item in
            guard let text = item.text?.trimmingCharacters(in: .whitespacesAndNewlines), !text.isEmpty else { return nil }
            let typeRaw = item.type ?? "action"
            let type = ExtractedItem.ItemType(rawValue: typeRaw) ?? .action
            return ExtractedItem(
                id: "ei\(Int(Date().timeIntervalSince1970 * 1000) + idx)",
                type: type,
                text: text,
                time: item.time ?? "",
                suggestedHour: item.suggestedHour ?? 8,
                status: "provisional"
            )
        }
    }
}

File: app/android/app/src/main/kotlin/com/mednote/service/AIService.kt
package com.mednote.service

import com.mednote.models.AppLanguage
import com.mednote.models.ExtractedItem
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.withContext

object AIService {
    private val client = OkHttpClient.Builder()
        .connectTimeout(60, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .build()

    data class PipelineResult(
        val transcriptLines: List<String>,
        val summary: String,
        val extractedItems: List<ExtractedItem>,
        val translatedTranscript: List<String>?
    )

    suspend fun runPipeline(
        audioFile: File,
        groqKey: String,
        geminiKey: String,
        userLanguage: AppLanguage,
        recordLanguage: AppLanguage
    ): PipelineResult = withContext(Dispatchers.IO) {
        coroutineScope {
            val gKey = groqKey.trim()
            val cKey = geminiKey.trim()
            val userLangName = userLanguage.label

            var originalText = "[No transcript — add a Groq API key in Settings]"
            if (gKey.isNotEmpty()) {
                groqWhisper(audioFile, gKey, recordLanguage.code)?.let { originalText = it }
            }

            val lines = splitTranscriptLines(originalText)
            val shouldTranslate = userLanguage != recordLanguage && !originalText.startsWith("[")

            val summaryDeferred = async {
                callGemini(
                    system = "Medical appointment summarizer. Warm, concise, plain language.",
                    message = """
                        Brief summary in $userLangName. Headings: 📋 Findings 💊 Medications ⚠️ Watch for 📅 Next steps

                        Transcript:
                        $originalText

                        Write entirely in $userLangName.
                    """.trimIndent(),
                    geminiKey = cKey,
                    groqKey = gKey
                )
            }

            val extractDeferred = async {
                callGemini(
                    system = "Medical data extractor. Return ONLY a valid JSON array, no markdown, no other text.",
                    message = """
                        Extract actionable items from the transcript. Respond ONLY with a valid JSON array (no markdown, no extra text), each item: {"type":"medication"|"appointment"|"action","text":"description in $userLangName","time":"time/date","suggestedHour":8}

                        IMPORTANT: All "text" values MUST be written in $userLangName.

                        Transcript:
                        $originalText
                    """.trimIndent(),
                    geminiKey = cKey,
                    groqKey = gKey
                )
            }

            val translateDeferred = if (shouldTranslate) {
                async {
                    callGemini(
                        system = "Medical transcript translator. Keep drug names, dosages, and numbers accurate. Return ONLY the translated transcript.",
                        message = "Translate into $userLangName:\n\n$originalText",
                        geminiKey = cKey,
                        groqKey = gKey
                    )?.let { splitTranscriptLines(it) }
                }
            } else null

            val summary = summaryDeferred.await()
                ?: "AI summary could not be generated — add a Groq or Gemini key in Settings."
            val extractRaw = extractDeferred.await()
            val translated = translateDeferred?.await()

            PipelineResult(
                transcriptLines = lines.ifEmpty { listOf(originalText) },
                summary = summary,
                extractedItems = parseExtractedItems(extractRaw),
                translatedTranscript = translated
            )
        }
    }

    private fun groqWhisper(file: File, apiKey: String, language: String): String? {
        val ext = file.extension.ifEmpty { "m4a" }
        val mime = if (ext == "wav") "audio/wav" else "audio/mp4"
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("file", "audio.$ext", file.asRequestBody(mime.toMediaType()))
            .addFormDataPart("model", "whisper-large-v3-turbo")
            .addFormDataPart("response_format", "verbose_json")
            .addFormDataPart("temperature", "0")
            .addFormDataPart("language", language)
            .build()

        val request = Request.Builder()
            .url("https://api.groq.com/openai/v1/audio/transcriptions")
            .header("Authorization", "Bearer $apiKey")
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return null
            val json = JSONObject(response.body?.string() ?: return null)
            return json.optString("text").ifEmpty { null }
        }
    }

    private fun callGemini(system: String, message: String, geminiKey: String, groqKey: String): String? {
        val prompt = "$system\n\n$message"
        if (geminiKey.isNotEmpty()) {
            for (model in listOf("gemini-2.5-flash", "gemini-2.5-flash-lite", "gemini-flash-latest")) {
                geminiGenerate(prompt, geminiKey, model)?.let { return it }
            }
        }
        if (groqKey.isNotEmpty()) {
            return groqChat(prompt, groqKey)
        }
        return null
    }

    private fun geminiGenerate(prompt: String, apiKey: String, model: String): String? {
        val url = "https://generativelanguage.googleapis.com/v1beta/models/$model:generateContent?key=$apiKey"
        val payload = JSONObject().apply {
            put("contents", org.json.JSONArray().put(
                JSONObject().put("role", "user").put("parts", org.json.JSONArray().put(
                    JSONObject().put("text", prompt)
                ))
            ))
            put("generationConfig", JSONObject()
                .put("maxOutputTokens", 1500)
                .put("temperature", 0.3))
        }
        val request = Request.Builder()
            .url(url)
            .post(payload.toString().toRequestBody("application/json".toMediaType()))
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return null
            val json = JSONObject(response.body?.string() ?: return null)
            if (json.has("error")) return null
            return json.optJSONArray("candidates")
                ?.optJSONObject(0)
                ?.optJSONObject("content")
                ?.optJSONArray("parts")
                ?.optJSONObject(0)
                ?.optString("text")
                ?.ifEmpty { null }
        }
    }

    private fun groqChat(prompt: String, apiKey: String): String? {
        val payload = JSONObject().apply {
            put("model", "llama-3.3-70b-versatile")
            put("messages", org.json.JSONArray().put(
                JSONObject().put("role", "user").put("content", prompt)
            ))
            put("max_tokens", 1500)
            put("temperature", 0.3)
        }
        val request = Request.Builder()
            .url("https://api.groq.com/openai/v1/chat/completions")
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .post(payload.toString().toRequestBody("application/json".toMediaType()))
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return null
            val json = JSONObject(response.body?.string() ?: return null)
            return json.optJSONArray("choices")
                ?.optJSONObject(0)
                ?.optJSONObject("message")
                ?.optString("content")
                ?.ifEmpty { null }
        }
    }

    private fun splitTranscriptLines(text: String): List<String> {
        if (text.isBlank()) return emptyList()
        val lines = mutableListOf<String>()
        Regex("[^.!?\\n]+[.!?\\n]*").findAll(text).forEach { match ->
            val s = match.value.trim()
            if (s.length > 1) lines.add(s)
        }
        if (lines.isEmpty()) {
            return text.lines().map { it.trim() }.filter { it.isNotEmpty() }
        }
        return lines
    }

    private fun parseExtractedItems(raw: String?): List<ExtractedItem> {
        if (raw.isNullOrBlank()) return emptyList()
        var cleaned = raw.replace("```json", "").replace("```", "")
        val start = cleaned.indexOf('[')
        val end = cleaned.lastIndexOf(']')
        if (start >= 0 && end > start) cleaned = cleaned.substring(start, end + 1)
        return runCatching {
            val array = Json.parseToJsonElement(cleaned).jsonArray
            array.mapIndexedNotNull { idx, element ->
                val obj = element.jsonObject
                val text = obj["text"]?.jsonPrimitive?.contentOrNull?.trim().orEmpty()
                if (text.isEmpty()) return@mapIndexedNotNull null
                ExtractedItem(
                    id = "ei${System.currentTimeMillis() + idx}",
                    type = obj["type"]?.jsonPrimitive?.contentOrNull ?: "action",
                    text = text,
                    time = obj["time"]?.jsonPrimitive?.contentOrNull ?: "",
                    suggestedHour = obj["suggestedHour"]?.jsonPrimitive?.intOrNull ?: 8,
                    status = "provisional"
                )
            }
        }.getOrDefault(emptyList())
    }
}
